pkg/ingester/shipper.go

```diff
@@ -0,0 +1,289 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+// Provenance-includes-location: https://github.com/thanos-io/thanos/blob/main/pkg/shipper/shipper.go
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Thanos Authors.
+
+// Package shipper detects directories on the local file system and uploads
+// them to a block storage.
+package ingester
+
+import (
+	"context"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/oklog/ulid"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/thanos-io/thanos/pkg/block"
+	"github.com/thanos-io/thanos/pkg/block/metadata"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/shipper"
+)
+
+type metrics struct {
+	dirSyncs        prometheus.Counter
+	dirSyncFailures prometheus.Counter
+	uploads         prometheus.Counter
+	uploadFailures  prometheus.Counter
+}
+
+func newMetrics(reg prometheus.Registerer) *metrics {
+	var m metrics
+
+	m.dirSyncs = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name: "thanos_shipper_dir_syncs_total",
+		Help: "Total number of dir syncs",
+	})
+	m.dirSyncFailures = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name: "thanos_shipper_dir_sync_failures_total",
+		Help: "Total number of failed dir syncs",
+	})
+	m.uploads = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name: "thanos_shipper_uploads_total",
+		Help: "Total number of uploaded blocks",
+	})
+	m.uploadFailures = promauto.With(reg).NewCounter(prometheus.CounterOpts{
+		Name: "thanos_shipper_upload_failures_total",
+		Help: "Total number of block upload failures",
+	})
+	return &m
+}
+
+// ThanosShipper watches a directory for matching files and directories and uploads
+// them to a remote data store.
+type ThanosShipper struct {
```

Comment:
I think naming it "ThanosShipper" should make it harder to understand for the future code reader. I would just name it "Shipper" (created by "NewShipper()"), given we're forking and we're going our way. The provenance is still respected using the `Provenance-includes-*` headers.

Reply:
Fair enough.